File: c/Problem010.c
// Problem 10
// By: Tim Oram
//
// Runtime: 1.791s

#include <stdio.h>

int main (void) {

	int primes[2000000] = {0};

	int i, j;
	
	unsigned long sum;

	// remove 2s
	for (i = 4; i <= 2000000; i += 2) {
		primes[i] = 1;
	}

	i = 3;

	while (i*i <= 2000000) {
		for (j = i+2; j < 2000000; j += 2) {
			if (!primes[j] && (j%i) == 0) {
				primes[j] = 1;
			}
		}
		
		do {
			i += 2;
		} while (primes[i]);
	}
	sum = 2;
	for (i = 3; i < 2000000; i += 2) {
		if (!primes[i]) {
			sum += i;
		}
	}

	printf("Answer: %ld\n", sum);
	return 0;
}

File: python/Problem010.py
# Problem 10
# By: Tim Oram


import math

num = 2000000

numbers = [True] * num

p = 3

numbers[0] = False
numbers[1] = False

for i in xrange(2, 2000000, 2):
	numbers[i] = False
	
while p*p <= num:
	for i in xrange(p+2, 2000000, 2):
		if numbers[i] and (i%p) == 0:
			numbers[i] = False
	while True:
		p += 2
		if numbers[p]:
			break

t = 2
for i in xrange(3, 2000000, 2):
	if numbers[i]:
		t += i
print "Answer:", t
